crates/block-producer/src/batch_builder/mod.rs

```diff
@@ -1,202 +1,195 @@
-}
+        loop {
+            tokio::select! {
+                _ = interval.tick() => {
+                    if inflight.len() >= self.workers.get() {
+                        tracing::info!("All batch workers occupied.");
+                        continue;
+                    }
+
+                    // Transactions available?
+                    let Some((batch_id, transactions)) =
+                        mempool.lock().await.select_batch()
+                    else {
+                        tracing::info!("No transactions available for batch.");
+                        continue;
+                    };
 
-impl<S, BB> DefaultBatchBuilder<S, BB>
-where
-    S: Store,
-    BB: BlockBuilder,
-{
-    // CONSTRUCTOR
-    // --------------------------------------------------------------------------------------------
-    /// Returns an new [BatchBuilder] instantiated with the provided [BlockBuilder] and the
-    /// specified options.
-    pub fn new(store: Arc<S>, block_builder: Arc<BB>, options: DefaultBatchBuilderOptions) -> Self {
-        Self {
-            store,
-            block_builder,
-            options,
-            ready_batches: Default::default(),
+                    inflight.spawn(batch_id, transactions);
+                },
+                result = inflight.join_next() => {
+                    let mut mempool = mempool.lock().await;
+                    match result {
+                        Err(err) => {
+                            tracing::warn!(%err, "Batch job panic'd.")
+                            // TODO: somehow embed the batch ID into the join error, though this doesn't seem possible?
+                            // mempool.batch_failed(batch_id);
+                        },
+                        Ok(Err((batch_id, err))) => {
+                            tracing::warn!(%batch_id, %err, "Batch job failed.");
+                            mempool.batch_failed(batch_id);
+                        },
+                        Ok(Ok((batch_id, batch))) => {
+                            mempool.batch_proved(batch_id, batch);
+                        }
+                    }
+                }
+            }
         }
     }
+}
 
-    // BATCH BUILDER STARTER
-    // --------------------------------------------------------------------------------------------
-    pub async fn run(self: Arc<Self>) {
-        let mut interval = time::interval(self.options.block_frequency);
+// BATCH WORKER
+// ================================================================================================
 
-        info!(target: COMPONENT, period_ms = interval.period().as_millis(), "Batch builder started");
+type BatchResult = Result<(BatchJobId, TransactionBatch), (BatchJobId, BuildBatchError)>;
 
-        loop {
-            interval.tick().await;
-            self.try_build_block().await;
+/// Wrapper around tokio's JoinSet that remains pending if the set is empty,
+/// instead of returning None.
+struct WorkerPool {
```

Comment:
Its possible to do it that way sure. My thinking was that `BatchBuilder` contains the configuration options which includes the worker count. Right now these always just default, but I imagine we would want to expose all these options as cli args e.g.
```
miden-node start block-producer --batch.interval ... --batch.workers 10 --batch.failure-rate 0.1
```
which would then be passed in here. If we move this around then its no longer pure configuration values, and instead also contains state which means you can no longer do things like trivially overriding defaults and instead need a bunch of `fn new`-like functions.

I'll move the worker count into the worker pool regardless; since that still works and probably makes more sense.